### PR TITLE
Guidelines for working with Memory<T> and Span<T>

### LINKED ISSUE #4823: Create usage guidelines for lifetime management of System.Memory<T>
Memory, and OwnedMemory are new types shipping in .NET Core 2.1.  Memory/OwnedMemory need  "rules of the road" usage guidance in order to write code that doesn't cause use-after-free bugs.

https://github.com/dotnet/corefx/issues/23863

https://github.com/dotnet/corefxlab/blob/master/docs/specs/span.md

https://github.com/dotnet/corefxlab/blob/master/docs/specs/memory.md

(The below is edited by @GrabYourPitchforks.)

The document at https://gist.github.com/GrabYourPitchforks/4c3e1935fd4d9fa2831dbfcab35dffc6 demonstrates how to think about `Memory<T>`. It goes over issues related to lifetime and gives a set of guidelines for developers who want to accept `Memory<T>` instances in their API surface. Read this document first.

The document at https://gist.github.com/GrabYourPitchforks/8efb15abbd90bc5b128f64981766e834 talks about `MemoryPool<T>` and related types, including how to write custom pools. It also goes over how to change the implementation of `Memory<T>` to get it to point to new categories of contiguous buffers, such as unmanaged memory. It covers more advanced scenarios.

### PR DESCRIPTION
**Pull request by rpetrusha**
## Guidelines for working with Memory\<T> and Span\<T>

Fixes #4823 

@ahsonkhan @stephentoub @GrabYourPitchforks

### CODE REVIEW

**Review by BillWagner — CHANGES_REQUESTED** [review]
This is really good @rpetrusha It's a clear explanation of these types and when to choose each.

I had a few suggestions for topics to introduce earlier that would make it more clear.

Overall, a great addition.

File: docs/standard/memory-and-spans/memory-t-usage-guidelines.md [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+
+## Owners, consumers, and lifetime management
+
+Since buffers can be passed around between APIs, and since buffers can sometimes be accessed from multiple threads, it is important to consider lifetime management. There are three core concepts:
+
+- **Ownership**. The owner of a buffer instance is responsible for lifetime management, including destroying the buffer when it is no longer in use. All buffers have a single owner. Generally the owner is the component that created the buffer or that received the buffer from a factory. Ownership can also be transferred; **Component-** can relinquish control of the buffer to **Component-B**, at which point **Component-A** may no longer use the buffer, and **Component-B** becomes responsible for destroying the buffer when it is no longer in use.
+
+- **Consumption**. The consumer of a buffer instance is allowed to use the buffer instance by reading from it and possibly writing to it. Buffers can have one consumer at a time unless some external synchronization mechanism is provided. Note that the active consumer of a buffer is not necessarily the buffer's owner.
```
[the remark this answers is not in the document]
Comment (rpetrusha) [reply docs/standard/memory-and-spans/memory-t-usage-guidelines.md]: Contractions don't seem to be my strong point, @BillWagner.

File: docs/standard/memory-and-spans/memory-t-usage-guidelines.md [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+
+The buffer has two consumers, `WriteInt32ToBuffer` and `DisplayBufferToConsole`. There is only one consumer at a time (first `WriteInt32ToBuffer`, then `DisplayBufferToConsole`), and neither of the consumers owns the buffer. Note also that "consumer" in this context does not imply a read-only view of the buffer; consumers can modify the buffer's contents, as `WriteInt32ToBuffer does, if given a read/write view of the buffer.
+
+The `WriteInt32ToBuffer` method has a lease on (can consume) the buffer between the start of the method call and the time the method returns. Similarly, `DisplayBufferToConsole` has a lease on the buffer while it is executing, and the lease is released when the method unwinds. (There is no API for lease management; a "lease" is simply a conceptual matter.)
+
+## Memory\<T> and the owner/consumer model
```
[the remark this answers is not in the document]
Comment (rpetrusha) [reply docs/standard/memory-and-spans/memory-t-usage-guidelines.md]: Thanks, @BillWagner; this is a really helpful comment. I've addressed it and committed. I'll address your following comment either later this evening or tomorrow.

**Review by BillWagner — APPROVED** [review]

**Comment by BillWagner** [comment]
I approved the changes, as I had no further issues @rpetrusha 

I'll let you work through with @mairaw and others if more needs to be done.

**Pull request merged**